dirstack/dirstack.1.in
.TH dirstack 1 "@RELEASE_DATE@" "@PACKAGE_STRING@" "User Commands"
.SH NAME
dirstack \- directory browser/selector
.SH SYNOPSIS
\fBdirstack\fP [ \fB-hnpq\fP ]
.SH DESCRIPTION
The \fBdirstack\fP utility is an interactive directory browser. It
maintains a `stack' of directories as do the \fBpushd\fP/\fBpopd\fP
commands of some shells, but in a visual, scrollable list.

When \fBdirstack\fP is run, it optionally adds the current working
directory to the top of the stack (if the \fB-p\fP switch is
supplied). If the current directory is already on the stack, it moves it
to the top of the stack. It then displays a \fIcurses\fP-based text
interface which shows the directory stack as a scrollable list.

A directory may be selected from the list with the keyboard, or--on
those terminals that support mouse input--with the mouse. The
following commands are recognized:
.TP 5
.B Down-Arrow or j
Any of these keys will move the highlight down to the next directory
in the list, scrolling the list if necessary.
.TP 5
.B Up-Arrow or k
Any of these keys will move the highlight up to the previous directory in the list, scrolling the list if necessary.
.TP 5
.B Right-Arrow, n, or Next-Page
Any of these keys will show the next screenful of directories.
.TP 5
.B Left-Arrow, p, or Prev-Page
Any of these keys will show the previous screenful of directories.
.TP 5
.B Backspace or d
Any of these keys will remove the currently-highlighted directory from the stack. If there is only one directory left on the stack, the command is ignored.
.TP 5
.B Return or Space
Any of these keys will accept the currently selected directory and exit the program.
.TP 5
.B ^L
Pressing this key will cause a full screen repaint.
.TP 5
.B q
This key will exit the program without accepting the selected directory.
.TP 0

On terminals that support mouse input, the mouse can be used to scroll
the list and select a directory. Single-clicking on a directory will
select the directory. Double-clicking on a directory will select and
accept the directory. Clicking on a ``--more--'' line has the same
effect as paging up or down in the corresponding direction.

\fBdirstack\fP must be integrated with the shell via shell aliases.

For \fBcsh/tcsh\fP the aliases are:

.RS 5
\fCalias ds 'dirstack && cd `cat ~/.dirstacksel`'

alias dspush 'dirstack -pn'
\fP
.RS -5

For \fBbash\fP the aliases are:

.RS 5
\fCfunction ds ()
  { dirstack && cd "`cat "$HOME/.dirstacksel"`"; }

alias dspush='dirstack -pn'
\fP
.RS -5

For \fBzsh\fP the aliases are:

.RS 5
\fCalias ds='dirstack && cd `cat ~/.dirstacksel`'

alias dspush='dirstack -pn'
\fP
.RS -5

For \fBksh\fP the aliases are:

.RS 5
\fCds ()
  { dirstack && cd "`cat "$HOME/.dirstacksel"`"; }

alias dspush='dirstack -pn'
\fP
.RS -5

Once the alias is defined, \fBdirstack\fP is used by invoking the
alias rather than by running the binary directly.

.SH OPTIONS
The following command line options are recognized:
.TP 5
.B -h
Display a command synopsis and copyright message, and exit.
.TP 5
.B -n
Do not start the browser (no stack display).
.TP 5
.B -p
Push the current directory onto the stack.
.TP 5
.B -q
Run quietly: do not beep the terminal in response to invalid keypresses.
.SH NOTES
There is presently no error checking to determine if any given
directory actually exists and/or is accessible.

\fBdirstack\fP will not run on a screen smaller than 40x10.
.SH FILES
.PD 0
.TP 20
.B ~/.dirstack
to save the contents of the directory stack

.PD 0
.TP 20
.B ~/.dirstacksel
to save the currently selected directory

.SH SEE ALSO
\fBcd(1)\fP
.SH AUTHOR
.PD 0
.TP 5
Mark Lindner <@PACKAGE_BUGREPORT@>


ChangeLog
2017-10-28  Mark Lindner  <markl@avalon>

	* stat/stat.c - fixed compiler warning
	* watcherd/* - removed
	* pfortune/* - removed
	* configure.ac - bumped version

2012-10-01  Mark Lindner  <markl@neuromancer>

	* aux-build/* - updated files
	* pkgenv/pkgenv.c - fixed compiler warnings

2011-06-26  Mark Lindner  <markl@neuromancer>

	* bin2c/* - added new utility
	* configure.ac, Makefile.am - added new bin2c utility, bump version

2008-04-23  Mark Lindner  <markl@neuromancer>

	* pfortune.1.in - fix typo in manpage

2008-04-21  Mark Lindner  <markl@neuromancer>

	* pkgenv.c - bugfix for shell handling

2008-04-20  Mark Lindner  <markl@neuromancer>

	* texi2ps, texi2ps.1.in - new script
	* pkgenv.c - bugfix: implement proper version string comparator
	* configure.ac - bump version

2007-12-10  Mark Lindner  <markl@neuromancer>

	* man2ps, man2ps.1.in, ps2pdf, ps2pdf.1.in, texi2pdf, texi2pdf.1.in,
	  man2pdf, man2pdf.1.in - modified scripts to accept multiple filename
	  arguments
	* <multiple> - rename 'tformat' utility to 'wrap'
	* pkgenv.c - fix memory crash bug

2007-12-08  Mark Lindner  <markl@neuromancer>

	* ranline.c - bugfix

2007-12-06  Mark Lindner  <markl@neuromancer>

	* ranline.c, ranline.1.in - add support for reading from stdin;
	  increase maximum supported line length to 1024

2007-11-19  Mark Lindner  <markl@neuromancer>

	* pkgenv.c, pkgenv.1.in - added -V (verbose) flag
	* basecvt.c, basecvt.1.in - added '?' command

2007-11-11  Mark Lindner  <markl@neuromancer>

	* pkgenv.c - improve error messages
	* pkgenv/automake_patch.txt - added

2007-09-16  Mark Lindner  <markl@neuromancer>

	* pkgenv.c, pkgenv.1.in - allow environment variable references in
	  variable substitution

2007-09-05  Mark Lindner  <markl@neuromancer>

	* pkgenv.c, pkgenv.1.in - more bugfixes & improvements

2007-09-01  Mark Lindner  <markl@neuromancer>

	* pkgenv.c - add special logic to handle old vs. new MANPATH dir;
	  don't add dirs that don't exist; bugfixes to 'del' command;
	  add DYLD_LIBRARY_PATH support for Mac OS X
	* *.1, *.1.in, configure.ac - generate manpages with correct version &
	  date in header
	* scripts/man2pdf - new script
	* Makefile.am - add 'pdf' target

2007-08-26  Mark Lindner  <markl@neuromancer>

	* <multiple> - code cleanup, bugfixes
	* pkgenv/, basecvt/ - new utilities added

2007-02-08  Mark Lindner  <markl@neuromancer>

	* scripts/Makefile.am - add manpages to dist
	* pfortune.c - fix segmentation fault

2006-04-17  Mark Lindner  <markl@neuromancer>

	* scripts/ - added useful shell scripts

2006-04-17  Mark Lindner  <markl@neuromancer>

	* dirstack.c, textlock.c, configure.ac - check for both ncurses.h
	  and ncurses/ncurses.h; fixes build error on OS X.

2006-04-15  Mark Lindner  <markl@neuromancer>

	* pfortune.c, pfortune.1 - partial rewrite, new -l switch
	* pfortune - added new databases (smileys, paul_graham, fortunes)

2006-04-13  Mark Lindner  <markl@neuromancer>

	* <multiple> - modified build files to locate libcfl using pkg-config
	* smileys.fdb - new fortune database added

2005-08-15  Mark Lindner  <markl@neuromancer>

	* dirstack.1, pascii.1 - manpage corrections
	* README - updated

2005-08-12  Mark Lindner  <markl@neuromancer>

	* <multiple> - Merged in 'djinn' project.
	* dirstack.c - Truncate paths that do not fit on one line.

2004-05-12  Mark Lindner  <markl@neuromancer>

	* bat.c - Added -8 switch.
	* <multiple> - Various minor code fixes.

2003-02-22  Mark Lindner  <markl@neuromancer>

	* <multiple> - Reorganized package so that there is only one
	  configure.ac (at the top level)
	* <multiple> - Modified all source code to use CFL library (which
	  replaces the pingutil library)
	* stat.c - don't exit with an error if getgrnam() or getpwent() fails,
	  but rather just display "?" for the user or group name
	* dirstack.c - fixed bug that created repaint glitch on delete,
	  added handling of ^L for screen refresh, and added -p switch.
	* ranline.c - use random() instead of rand() so that (on Solaris, at
	  least) we're not limited to a random number range of 32767.
	* <multiple> - updated manpages
	* pascii.c - handle DEL key and high-bit chars more gracefully
	* textlock.c - use C_system_get_hostname() to get hostname; the old way
	  of copying the results of uname() was non-portable, because SYS_NMLN
	  is not defined on some systems (namely OS X).

2003-01-02  Mark Lindner  <markl@neuromancer>

	* <multiple> - replaced `bool_t' with `p_bool_t' to fix compilation
	  errors when linked against the latest pingutil

2002-04-17  Mark Lindner  <mark_a_lindner@yahoo.com>

	* <multiple> - reorganized the entire source tree to use automake &
	  autoconf. Damn this was a lot of work (especially since I'd never
	  used automake before). Also made several minor fixes to make the
	  packages compile cleanly on Linux.

2002-04-14  Mark Lindner  <mark_a_lindner@yahoo.com>

	* dirstack - new utility

2001-10-14  Mark Lindner  <mark_a_lindner@yahoo.com>

	* <multiple> - minor modifications for pingutil-2.2
	* watcherd - new utility
	* pfortune - new utility

2001-09-14  Mark Lindner  <mark_a_lindner@yahoo.com>

	* <multiple> - slight modifications to use pingutil-2.1
	* ptree - renamed to dirtree to avoid confusion with Solaris ptree cmd
	* dirtree - fixed segmentation fault


dirstack/dirstack.c
/* ----------------------------------------------------------------------------
   dirstack - directory browser/selector
   Copyright (C) 1994-2011  Mark A Lindner

   This file is part of misctools.

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, see
   <http://www.gnu.org/licenses/>.
   ----------------------------------------------------------------------------
*/

/* --- Feature Test Switches --- */

#include "config.h"

/* --- System Headers --- */

#include <stdio.h>
#if defined HAVE_NCURSES_NCURSES_H
#include <ncurses/ncurses.h>
#elif defined HAVE_NCURSES_H
#include <ncurses.h>
#endif
#include <string.h>
#include <ctype.h>
#include <signal.h>
#include <termios.h>
#include <sys/ioctl.h>

#include <cbase/cbase.h>

/* --- Local Headers --- */

/* --- Macros --- */

#define HEADER "dirstack v" VERSION " - Mark Lindner"
#define USAGE "[ -hnpq ]"

#define center(S)                               \
  ((COLS - strlen(S)) / 2)

#define DATA_FILE ".dirstack"
#define DATA_SELFILE ".dirstacksel"

#define MIN_COLS 30
#define MIN_LINES 10

/* --- File Scope Variables --- */

static const char *title = "dirstack " VERSION;
static const char *more[2] = { "        ", "--more--" };
static WINDOW *mainwin, *scrollwin, *infowin;
static char *data_file, *data_selfile, *homedir, **dirlist;
static const char *newdir;
static int scrollsz, top = 0, selpos = 0;
static size_t listlen = 0;
static c_bool_t quiet = FALSE, pushcurrent = FALSE;

/* --- Functions --- */

static void redraw(void);

static void buildscreen(void)
{
  cbreak();
  noecho();

  if((LINES < MIN_LINES) || (COLS < MIN_COLS))
  {
    endwin();
    C_error_printf("Screen is too small. Must be at least %dx%d.\n",
                   MIN_COLS, MIN_LINES);
    exit(EXIT_FAILURE);
  }

  scrollsz = LINES - 6;

#ifdef NCURSES_MOUSE_VERSION
  mousemask((BUTTON1_CLICKED | BUTTON1_DOUBLE_CLICKED), NULL);
#endif

  /* draw screen */

  mainwin = newwin(LINES, COLS, 0, 0);

  scrollwin = derwin(mainwin, scrollsz, COLS, 2, 0);
  scrollok(scrollwin, TRUE);

  infowin = derwin(mainwin, 2, COLS, LINES - 2, 0);
  keypad(infowin, TRUE);
  noqiflush();
}

/*
 */

static void flow_text(WINDOW *win, const char *text, int cols, int rows)
{
  int n = 0, y = 0;
  const char *word, *ctx;
  size_t len;

  wmove(win, 0, 0);

  for(word = C_string_tokenize(text, " \n", &ctx, &len);
      word;
      word = C_string_tokenize(NULL, " \n", &ctx, &len))
  {
    if((n + len + 1) > cols)
    {
      if(++y == rows)
        return; // full

      wmove(win, y, 0);
    }

    waddnstr(win, word, len);
    waddch(win, ' ');

    n += (len + 1);
  }
}

/*
 */

static void drawscreen(void)
{
  int i;
  size_t nl = strlen(newdir);
  size_t ml = COLS - 3;
  c_bool_t truncated = FALSE;

  wmove(mainwin, 0, 0);
  wattron(mainwin, A_REVERSE);
  for(i = 0; i < COLS; i++)
  {
    mvwaddch(mainwin, 0, i, ' ');
    mvwaddch(mainwin, LINES - 3, i, ' ');
  }
  mvwaddstr(mainwin, 0, center(title), title);

  if(nl > ml)
  {
    truncated = TRUE;
    ml -= 3;
  }

  mvwaddnstr(mainwin, LINES - 3, 1, newdir, ml);
  if(truncated)
    waddstr(mainwin, "...");

  wattroff(mainwin, A_REVERSE);

  flow_text(infowin, "Use arrows or mouse click to select,"
            " <Return> or double-click to accept, <Backspace> to"
            " delete, <q> to quit.", COLS, LINES);
}

/*
 */

static void adjust(int sig)
{
  struct winsize size;

  if(ioctl(STDOUT_FILENO, TIOCGWINSZ, &size) == 0)
  {
    LINES = size.ws_row;
    COLS = size.ws_col;
  }

  endwin();
  doupdate();

  buildscreen();
  drawscreen();
  redraw();

  wrefresh(mainwin);
  wrefresh(infowin);

  /* reposition cursor */
  wmove(infowin, 1, COLS - 1);

  /* Flush input. Why the hell is this necessary? if we don't, the
   * next character typed is "lost". I don't get it.
   */

  nodelay(infowin, TRUE);
  while(wgetch(infowin) != ERR) { };
  nodelay(infowin, FALSE);

}

/*
 */

static void drawitem(int i)
{
  char *s = dirlist[i];
  int l = strlen(homedir);
  int ml = COLS - 3;
  size_t sl = strlen(s);
  c_bool_t truncated = FALSE;

  if(!strncmp(s, homedir, l))
  {
    waddstr(scrollwin, "~");
    s += l;
    sl -= l;
    sl++;
  }

  if(sl > ml)
  {
    truncated = TRUE;
    ml -= 3;
  }

  waddnstr(scrollwin, s, ml);
  if(truncated)
    waddstr(scrollwin, "...");
}

/*
 */

static void dehighlight(void)
{
  mvwaddstr(scrollwin, selpos, 0, " ");
  drawitem(top + selpos);
  waddstr(scrollwin, " ");
}

/*
 */

static void highlight(void)
{
  wattron(scrollwin, A_REVERSE);
  dehighlight();
  wattroff(scrollwin, A_REVERSE);
}

/*
 */

static void updatemore(void)
{
  wattron(mainwin, A_BOLD);
  mvwaddstr(mainwin, 1, 0, more[top > 0]);
  mvwaddstr(mainwin, scrollsz + 2, 0, more[(top + scrollsz) < listlen]);
  wattroff(mainwin, A_BOLD);
  wrefresh(mainwin);
}

/*
 */

static void redraw(void)
{
  char **p;
  int i, z;

  werase(scrollwin);
  for(i = 0, z = top, p = dirlist + top; (i < scrollsz) && *p; i++, p++, z++)
  {
    wmove(scrollwin, i, 1);
    drawitem(z);
  }

  highlight();
  updatemore();

  wrefresh(scrollwin);

}

/*
 */

static void moveup(void)
{
  if((top + selpos) > 0)
  {
    dehighlight();

    if(selpos == 0)
    {
      wmove(scrollwin, 0, 0);
      winsertln(scrollwin);
      top--;
    }
    else
      selpos--;

    highlight();
    updatemore();

    wrefresh(scrollwin);
  }
  else
    if(! quiet)
      beep();
}

/*
 */

static void pageup(void)
{
  if(top >= scrollsz)
  {
    selpos = 0;
    top -= scrollsz;
    redraw();
  }
  else if(top > 0)
  {
    selpos = top = 0;
    redraw();
  }
  else
    if(! quiet)
      beep();
}

/*
 */

static void movedown(void)
{
  if((top + selpos) < (listlen - 1))
  {
    dehighlight();

    if(selpos == (scrollsz - 1))
    {
      wmove(scrollwin, 0, 0);
      wdeleteln(scrollwin);
      top++;
    }
    else
      selpos++;

    highlight();
    updatemore();

    wrefresh(scrollwin);
  }
  else
    if(! quiet)
      beep();
}

/*
 */

static void pagedown(void)
{
  if((top + scrollsz) < listlen)
  {
    selpos = 0;
    top += scrollsz;
    redraw();
  }

  else
    if(! quiet)
      beep();
}

/*
 */

static void delete(void)
{
  char **q, **p;

  if(listlen > 1)
  {
    for(p = (dirlist + top + selpos), q = (p + 1); *p; p++, q++)
      *p = *q;

    if((top + selpos) == (listlen - 1))
      selpos--;

    listlen--;

    redraw();
  }
  else
    if(! quiet)
      beep();
}

/*
 */

static c_bool_t dirstack_write(void)
{
  FILE *fp;
  char **p, **sel;

  if(!(fp = fopen(data_selfile, "w+")))
    return(FALSE);

  fputs(dirlist[top + selpos], fp);

  fputc('\n', fp);
  fclose(fp);

  /* write the selected one first, then the rest */

  if(!(fp = fopen(data_file, "w+")))
    return(FALSE);

  fputs(dirlist[top + selpos], fp);
  fputc('\n', fp);

  sel = (dirlist + top + selpos);
  for(p = dirlist; *p; p++)
  {
    if(p != sel)
    {
      fputs(*p, fp);
      fputc('\n', fp);
    }
  }
  fclose(fp);

  return(TRUE);
}

/*
 */

static char ** dirstack_read(size_t *listlen)
{
  FILE *fp;
  char buf[1024];
  c_vector_t *vec;
  size_t c = 0;

  vec = C_vector_start(20);

  if(pushcurrent)
  {
    ++c;
    C_vector_store(vec, C_string_dup(newdir));
  }

  if((fp = fopen(data_file, "r")))
  {
    while(C_io_gets(fp, buf, sizeof(buf), '\n') != EOF)
    {
      int l;

      if(! *buf)
        continue;

      l = strlen(buf) - 1;
      if((l > 0) && (*(buf + l) == '/'))
        *(buf + l) = NUL;

      if(isspace((int)*buf))
        continue;

      if(!strcmp(buf, newdir) && pushcurrent)
        continue; /* it was already pushed */

      C_vector_store(vec, C_string_dup(buf));
      ++c;
    }

    fclose(fp);
  }

  if(c == 0)
    C_vector_store(vec, C_string_dup(newdir));

  return(C_vector_end(vec, listlen));
}

/*
 */

int main(int argc, char **argv)
{
  c_bool_t cancel = FALSE, errflag = FALSE, nodisplay = FALSE;
  int c;

  C_error_init(*argv);

  /* parse the command line */

  while((c = getopt(argc, argv, "hnpq")) != EOF)
  {
    switch(c)
    {
      case 'h':
        C_error_printf("%s\n", HEADER);
        C_error_usage(USAGE);
        exit(EXIT_SUCCESS);
        break;

      case 'p':
        pushcurrent = TRUE;
        break;

      case 'n':
        nodisplay = TRUE;
        break;

      case 'q':
        quiet = TRUE;
        break;

      default:
        errflag = TRUE;
        break;
    }
  }

  /* catch illegal option errors */

  if(errflag)
  {
    C_error_usage(USAGE);
    exit(EXIT_FAILURE);
  }

  signal(SIGWINCH, adjust);

  if(!(newdir = C_file_getcwd()))
  {
    C_error_printf("Unable to determine working directory.\n");
    exit(EXIT_FAILURE);
  }

  /* initialize paths */

  if(!(homedir = C_system_get_homedir()))
  {
    C_error_printf("Unable to determine home directory.\n");
    exit(EXIT_FAILURE);
  }

  data_file = C_newstr(strlen(homedir) + strlen(DATA_FILE) + 2);
  strcpy(data_file, homedir);
  strcat(data_file, "/");
  strcat(data_file, DATA_FILE);

  data_selfile = C_newstr(strlen(homedir) + strlen(DATA_SELFILE) + 2);
  strcpy(data_selfile, homedir);
  strcat(data_selfile, "/");
  strcat(data_selfile, DATA_SELFILE);

  /* read in directory list */

  if(!(dirlist = dirstack_read(&listlen)))
  {
    C_error_printf("Error reading directory stack file.\n");
    exit(EXIT_FAILURE);
  }

  if(nodisplay)
  {
    if(!dirstack_write())
    {
      C_error_printf("Error writing directory stack file(s).\n");
      exit(EXIT_FAILURE);
    }

    exit(EXIT_SUCCESS);
  }

  /* init display */

  initscr();
  buildscreen();
  drawscreen();
  wrefresh(mainwin);
  redraw();

  /* command processing loop */

  for(;;)
  {
    int c;

    c = mvwgetch(infowin, 1, COLS - 1);
    switch(c)
    {
      case 'k':
      case KEY_UP:
        moveup();
        break;

      case 'j':
      case KEY_DOWN:
        movedown();
        break;

      case 'd':
      case KEY_BACKSPACE:
        delete();
        break;

      case 'n':
      case KEY_NPAGE:
      case KEY_RIGHT:
        pagedown();
        break;

      case 'p':
      case KEY_PPAGE:
      case KEY_LEFT:
        pageup();
        break;

      case KEY_ENTER:
      case '\n':
      case ' ':
        goto END;
        break;

      case 'q':
        cancel = TRUE;
        goto END;

      case 12:
        wrefresh(curscr);
        break;

#ifdef NCURSES_MOUSE_VERSION
      case KEY_MOUSE:
      {
        MEVENT mevt;
        int y;

        getmouse(&mevt);
        y = mevt.y;

        if(y == 1)
          pageup();

        else if(y == scrollsz + 2)
          pagedown();

        y -= 2;
        if(y >= 0 && ((y + top) < listlen))
        {
          dehighlight();
          selpos = y;
          highlight();
          wrefresh(scrollwin);

          if(mevt.bstate & BUTTON1_DOUBLE_CLICKED)
          {
            usleep(100000);
            goto END;
          }
        }

        break;
      }
#endif

      default:
        if(! quiet)
          beep();
    }
  }

  /* clean up & exit */

END:
  endwin();

  if(!cancel)
    if(!dirstack_write())
    {
      C_error_printf("Error writing directory stack file(s).\n");
      exit(EXIT_FAILURE);
    }

  exit(cancel ? EXIT_FAILURE : EXIT_SUCCESS);
}

/* end of source file */
